Introduce cindent cinoptions entry to treat pragmas as normal code

Maxr1998 opened the pull request:
The new option `cinpragma` or `cinp` allows pragma directives, e.g. for OpenMP,
to be treated as normal code, and have indented like the rest of the
program.

Compare the following two examples:

The first is formatted with the current implementation, or with `cinpragma` set to `"false"`. As you can see, the `#pragma` directives are treated like any other directive, and moved to the start of the line.

```c
int fib(int n) {
    int i, j;

    if (n < 2) { 
        return n;
    } else {
#pragma omp parallel
        {
#pragma omp single
            {
#pragma omp task shared(i)
                i = fib(n - 1);
#pragma omp task shared(j)
                j = fib(n - 2);
#pragma omp taskwait
            }
        }
        return i + j;
    }
}
```

In the second example, `cinpragma` is set to `"true"`. Note how the pragmas are now part of the normal code, and indented like it as well. This highly improves the readability _in that case_, and thus makes a lot of sense for such OpenMP code.

```c
int fib(int n) {
    int i, j;

    if (n < 2) { 
        return n;
    } else {
        #pragma omp parallel
        {
            #pragma omp single
            {
                #pragma omp task shared(i)
                i = fib(n - 1);
                #pragma omp task shared(j)
                j = fib(n - 2);
                #pragma omp taskwait
            }
        }
        return i + j;
    }
}
```

This PR adds the aforementioned option, which is off by default though, to be as little intrusive as possible. What's obviously still missing is some documentation for the option, and possibly a better name for it. Also, I'm not 100% sure that my codestyle is right, so I'll gladly take any feedback and suggestions for improvement.

brammool commented:
Since this changes the indent of a specific item, it is more appropriate to add an entry in 'cinoptions'.  '#' is already taken, 'P' could be used.

Maxr1998 commented:
That should be better now :smile:

Maxr1998 commented:
I've confirmed it working by adding `Ps` to `cinoptions`.

k-takata commented on runtime/doc/indent.txt:
```diff
@@ -570,9 +570,15 @@ The examples below assume a 'shiftwidth' of 4.
 	      with "#" does not work.
 
 
+	PN    When N is non-zero recognize C pragmas, does not concern other
```
I think it is unclear that it treats pragmas as normal code.

Maxr1998 replied:
Yeah, I improved it now.

k-takata commented on src/cindent.c:
```diff
@@ -2114,8 +2118,10 @@ get_c_indent(void)
+    // #defines and so on go at the left when included in 'cinkeys',
+    // exluding pragmas when customized in 'cinoptions'
+    if (*theline == '#' && (*linecopy == '#' || in_cinkeys('#', ' ', TRUE))
+		&& (curbuf->b_ind_pragma == 0 || STRNCMP(theline, "#pragma", 7) != 0))
```
This doesn't recognize `# pragma` (space between `#` and `pragma`).

Maxr1998 replied:
Fixed.

k-takata reviewed (commented):
It might be better to add a test in `src/testdir/test_cindent.vim`.

Maxr1998 commented:
@brammool could you take another look, please? It should be good now.

brammool commented:
A quick look shows it should be OK.  It's in the todo list.

Maxr1998 commented:
Thank you :blush: There's no hurry. Just wanted to make sure my PR isn't forgotten :wink:

Maxr1998 commented:
@brammool I just rebased this onto the latest upstream, and combined it into a single commit, could you take another look at this? Would love to get this finally in.

Maxr1998 committed:
cindent: introduce cinoptions entry to treat pragmas like normal code

The new entry P in 'cinoptions' allows pragma directives (e.g. for OpenMP)
to be treated as normal code, and have them indented like the rest of the
program.
